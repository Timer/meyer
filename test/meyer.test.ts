import path from 'path';
import sqlite3 from 'sqlite3';
import { afterAll, beforeAll, describe, expect, test } from 'vitest';
import Meyer from '../src';
import SqliteDbms from './SqliteDbms';

const noop = () => {};

describe('Meyer', () => {
  let meyer: Meyer;
  let db!: sqlite3.Database;
  let dbms!: SqliteDbms;
  beforeAll(() => {
    db = new sqlite3.Database(':memory:');
    dbms = new SqliteDbms(db);
  });

  test('can execute with nothing', async () => {
    meyer = new Meyer({
      migrationsPath: path.join(__dirname, './migrations'),
      dbms,
      development: true,
    });

    await meyer.execute();
  });

  test('creates an example table (dev)', async () => {
    meyer = new Meyer({
      migrationsPath: path.join(__dirname, './migrations-01'),
      dbms,
      development: true,
    });

    await meyer.execute();
    expect(await dbms.all('SELECT * FROM "example";')).toMatchSnapshot();
  });

  test('creates an example table (prod)', async () => {
    meyer = new Meyer({
      migrationsPath: path.join(__dirname, './migrations-01-prod'),
      dbms,
    });

    await meyer.execute();
    expect(await dbms.all('SELECT * FROM "example2";')).toMatchSnapshot();
  });

  test('can update the example table', async () => {
    meyer = new Meyer({
      migrationsPath: path.join(__dirname, './migrations-01-v2'),
      dbms: dbms,
      development: true,
    });

    await meyer.execute();
    expect(await dbms.all('SELECT * FROM "example";')).toMatchSnapshot();
  });

  test('throws an error when changing non-development', async () => {
    meyer = new Meyer({
      migrationsPath: path.join(__dirname, './migrations-01'),
      dbms,
      development: true,
    });

    await meyer.execute();
    expect(await dbms.all('SELECT * FROM "example";')).toMatchSnapshot();

    const changeNonDevelopment = async () => {
      meyer = new Meyer({
        migrationsPath: path.join(__dirname, './migrations-01-v2'),
        dbms,
        development: false,
      });
      await meyer.execute();
    };

    let err;
    await changeNonDevelopment().then(noop, (e) => (err = e));
    expect(err).toMatchSnapshot();
  });

  test('throws an error when up is missing', async () => {
    meyer = new Meyer({
      migrationsPath: path.join(__dirname, './migrations-no-up'),
      dbms,
    });

    let err;
    await meyer.execute().then(noop, (e) => (err = e));
    expect(err).toMatchSnapshot();
  });

  test('throws an error when down is missing', async () => {
    meyer = new Meyer({
      migrationsPath: path.join(__dirname, './migrations-no-down'),
      dbms,
    });

    let err;
    await meyer.execute().then(noop, (e) => (err = e));
    expect(err).toMatchSnapshot();
  });

  afterAll(() => {
    db.close();
  });
});
