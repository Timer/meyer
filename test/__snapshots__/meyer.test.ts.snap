// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Meyer > can update the example table 1`] = `
[
  {
    "id": 1,
    "payload": "foo",
  },
  {
    "id": 2,
    "payload": "bar",
  },
  {
    "id": 3,
    "payload": "baz",
  },
]
`;

exports[`Meyer > creates an example table (dev) 1`] = `
[
  {
    "id": 1,
    "payload": "foo",
  },
  {
    "id": 2,
    "payload": "bar",
  },
]
`;

exports[`Meyer > creates an example table (prod) 1`] = `
[
  {
    "id": 1,
    "payload": "foo",
  },
  {
    "id": 2,
    "payload": "bar",
  },
]
`;

exports[`Meyer > throws an error when changing non-development 1`] = `
[
  {
    "id": 1,
    "payload": "foo",
  },
  {
    "id": 2,
    "payload": "bar",
  },
]
`;

exports[`Meyer > throws an error when changing non-development 2`] = `[Error: ... migration 1 has a hash mismatch, aborting ...]`;

exports[`Meyer > throws an error when down is missing 1`] = `[Error: no down in migration 1 (example)]`;

exports[`Meyer > throws an error when up is missing 1`] = `[Error: no up in migration 1 (example)]`;
